pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import annotations
+
+from typing import (
+    Iterable,
+    cast,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    FrameOrSeries,
+    PositionalIndexer,
+)
+from pandas.util._decorators import (
+    cache_readonly,
+    doc,
+)
+
+from pandas.core.dtypes.common import (
+    is_integer,
+    is_list_like,
+)
+
+from pandas.core.groupby import groupby
+from pandas.core.indexes.api import CategoricalIndex
+
+
+class GroupByIndexingMixin:
+    """
+    Mixin for adding .rows to GroupBy.
+    """
+
+    @property
+    def rows(self) -> _rowsGroupByIndexer:
+        return _rowsGroupByIndexer(cast(groupby.GroupBy, self))
+
+
+@doc(GroupByIndexingMixin.rows)
+class _rowsGroupByIndexer:
+    def __init__(self, grouped: groupby.GroupBy):
+        self.grouped = grouped
+
+    def __getitem__(self, arg: PositionalIndexer | tuple) -> FrameOrSeries:
+        with groupby.group_selection_context(self.grouped):
```

Comment:
if this is the reason you need `groupby` then we have to move that somewhere else or the imports are convoluted.

Reply:
I also needed it in line 35 `return _rowsGroupByIndexer(cast(groupby.GroupBy, self))`
The class is a mixin but I had to cast it to the "mixed-in" class since I needed access to GroupBy attributes and methods.
I copied the code style from DataFrame.iloc, but since that does not use types, so does not have the problem:
`class _iLocIndexer(_LocationIndexer):`

Reply:
for the `cast` bits you can put the import inside a `if TYPE_CHECKING:` block